drivers: adc_mcux: improve initialization of the ADC

jfischer-no opened the pull request:
* save pointer for repeated sampling
* improve initialization of the ADC

jfischer-no committed:
drivers: adc_mcux: save pointer for repeated sampling

Save pointer for repeated sampling.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
drivers: adc_mcux: improve initialization of the ADC

Improve initialization of the ADC.

After initialization with the default values, the clock source
of the ADC is asynchronous clock (ADACK) and clock divide ratio
is 8. The minimum conversion clock frequency is 1MHz.
Add clock divider selection and set default divide ratio to 1.
That sets the conversion clock frequency to approximately 5MHz.

Default configuration for the voltage reference is set to
external pins V_REFH and V_REFL. Depending on the MCU configuration
V_REFL may be connected to ground and V_REFH to VREF_OUT.
Since Voltage Reference block is not supported, the ADC does not work
properly on FRDM-KW41Z. Add voltage reference selection to fix it.

Enable self-calibration function as recommeded in Reference Manual.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

anangl commented on drivers/adc/adc_mcux_adc16.c:
```diff
@@ -180,6 +180,7 @@ static void mcux_adc16_isr(void *arg)
 	SYS_LOG_DBG("Finished channel %d. Result is 0x%04x",
 		    data->channel_id, result);
 
+	data->repeat_buffer = data->buffer;
```
This should not be done here but in `adc_context_start_sampling()` instead. When a sampling is repeated, all channels selected for the sequence are sampled again, thus `data->repeat_buffer` should point to the result from the first sampled channel, and here `data->buffer` points to the last one (so this will work only for single channel samplings).

After moving this assignment, [the one from `start_read()`](https://github.com/zephyrproject-rtos/zephyr/blob/5a098d2b8977c7d24d5d56b4c80a8dcc8cf83a47/drivers/adc/adc_mcux_adc16.c#L99) will become redundant.

MaureenHelm commented on drivers/adc/Kconfig.mcux:
```diff
@@ -11,3 +11,37 @@ config ADC_MCUX_ADC16
+if ADC_MCUX_ADC16
+
+choice
+	prompt "Clock Divide Selection"
```
The clock divider and voltage reference should probably come from device tree instead.

jfischer-phytec-iot replied:
That was my first thought. Kconfig was then more convenient to try out another configuration. Also, the settings could differ from sample to sample.

jfischer-no committed:
boards: frdm_kw41z: set VREF to alternate source

Since Voltage Reference block is not supported, the ADC does not work
properly on FRDM-KW41Z. Set voltage reference to alternate source
in frdm_kw41z_defconfig.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-phytec-iot commented on tests/drivers/adc/adc_api/boards/frdm_kw41z.conf:
```diff
@@ -0,0 +1 @@
+CONFIG_ADC_MCUX_ADC16_VREF_ALTERNATE=y
```
Done. I also tried the default reference. 
```c
static int mcux_adc16_init_vref(struct device *dev)                                                      
{                                                                                                        
        vref_config_t vref_config;                                                                       
                                                                                                         
        if (!FSL_FEATURE_SOC_VREF_COUNT) {                                                               
                SYS_LOG_ERR("Error");                                                                    
                return -EIO;                                                                             
        }                                                                                                
                                                                                                         
        CLOCK_EnableClock(kCLOCK_Vref0);                                                                 
        VREF_GetDefaultConfig(&vref_config);                                                             
        VREF_Init((void*)0x40074000, &vref_config);                                                      
                                                                                                         
        return 0;                                                                                        
}                      
```
It works, but I think this should be done in something like voltage/current regulators driver.
